### PR TITLE
mm/gran: remove size limitation

### PR DESCRIPTION
**Pull request by yf13**
## Summary

Current granule supports allocating at most 32 granules, which is inconvenient when large spaces are needed (e.g. for mapping a frame buffer).  This patch removes the limitation with the help of a `gran_range_s` structure for granule allocation table operations.

major changes from current mainline:
- The range size handling details moved to `mm_grantable.c`, or more specifically the `gran_set_()` and `gran_match()` functions. 
- Simplification of `gran_alloc()`,`gran_free()` and `gran_reserve()` as they no longer care sizing details.

So to see how this patch works, better look at `gran_match()` and `gran_set_()` functions in `mm_grantable.c`. The `gran_search()` relies on the last mismatch position returned by `gran_match()` to speed up free space searching.

## Impact
None

## Testing

- CI checks
- 12.5.1-rc0 `rv-virt/knsh32` plus this patch
- 12.5.1-rc0 `rv-virt/knsh64` plus this patch
- 12.5.1-rc0 `canmv230/nsbi` plus this patch

### CODE REVIEW

**Comment by tmedicci** [comment]
Just curious, why not remove the limitation from the current gran allocator?

**Comment by yf13** [comment]
> Just curious, why not remove the limitation from the current gran allocator?

I don't want break things. Actually `gran_alloc2()`, `gran_free2()` and `gran_reserve2()` can be viewed as a rewrite of current alloc/free/reserve functions. They share the same `struct gran_s`.

**Comment by tmedicci** [comment]
> > Just curious, why not remove the limitation from the current gran allocator?
> 
> I don't want break things. Actually `gran_alloc2()`, `gran_free2()` and `gran_reserve2()` can be viewed as a rewrite of current alloc/free/reserve functions. They share the same `struct gran_s`.

I don't think it would break things in such case. If it just removes the limitation, I think we can just upgrade it to avoid any confusion (and to improve maintainability, of course).

**Comment by acassis** [comment]
> > > Just curious, why not remove the limitation from the current gran allocator?
> > 
> > 
> > I don't want break things. Actually `gran_alloc2()`, `gran_free2()` and `gran_reserve2()` can be viewed as a rewrite of current alloc/free/reserve functions. They share the same `struct gran_s`.
> 
> I don't think it would break things in such case. If it just removes the limitation, I think we can just upgrade it to avoid any confusion (and to improve maintainability, of course).

@tmedicci I understand @yf13 because this improvement needs to be well tested. Maybe we could use same logic that was used when we replaced our legacy printf() function with some smaller one developed by Keith Packard: he added his new libc but added an option to use the older.

**Comment by tmedicci** [comment]
> > > > Just curious, why not remove the limitation from the current gran allocator?
> > > 
> > > 
> > > I don't want break things. Actually `gran_alloc2()`, `gran_free2()` and `gran_reserve2()` can be viewed as a rewrite of current alloc/free/reserve functions. They share the same `struct gran_s`.
> > 
> > 
> > I don't think it would break things in such case. If it just removes the limitation, I think we can just upgrade it to avoid any confusion (and to improve maintainability, of course).
> 
> @tmedicci I understand @yf13 because this improvement needs to be well tested. Maybe we could use same logic that was used when we replaced our legacy printf() function with some smaller one developed by Keith Packard: he added his new libc but added an option to use the older.

That makes sense. Use version 2 as the default and add an option to enable the legacy version. If something fails with the new version, we would know soon...

**Comment by xiaoxiang781216** [comment]
but, it's hard to compare the change with the duplication.

**Comment by tmedicci** [comment]
> but, it's hard to compare the change with the duplication.

yeah, adding a new version makes it more difficult to review the changes. Maybe we can try to substitute the granule allocator (suggestion, do it in a separate PR and let this as a draft if we need it in the future to stick with the two versions). The updated version of the granule allocator can be reverted if it breaks something.

**Comment by yf13** [comment]
> But it's hard to compare the change with the duplication. Why not add test to verify the new code work as expect? The duplication doesn't improve the code magically, instead hiding the bug since it disables by default.

I don't have a "complete" test set yet, if there is such a test set, please teach. 
On the other side, it seems working with rv64 qemu, but the rv32 qemu is broken with the current allocator so I can't finish checking.

**Comment by yf13** [comment]
@xiaoxiang781216  and @acassis , I checked this patch with  `rv-virt/knsh32`, `rv-virt/knsh64` and `canmv230/nsbi` and the `12.5.1-rc0` version. The `ostest` result is same as original allocator, I also checked the details with simple testing function and both old and new allocators are fine with the same test set.

So please let me know how to proceed.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 and @acassis , I checked this patch with `rv-virt/knsh32`, `rv-virt/knsh64` and `canmv230/nsbi` and the `12.5.1-rc0` version, though `rv-virt/knsh32` has [issue#12123](https://github.com/apache/nuttx/issues/12123). The `ostest` result is same as original allocator, I also checked details with a simple test function and both old and new allocators are fine.
> 
> So please let me know how to proceed.

Thank for testing, let's improve gran inplace.

**Comment by yf13** [comment]
@xiaoxiang781216 the replacement is done, please review.

**Comment by yf13** [comment]
> Thank for testing, let's improve gran in place.
@xiaoxiang781216 replacement is done.

File: mm/mm_gran/mm_gran2.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
+/****************************************************************************
+ * mm/mm_gran/mm_gran2.c
```
Comment (xiaoxiang781216) [inline mm/mm_gran/mm_gran2.c:2]: why mm_gran2.c

**Review by acassis — APPROVED** [review]

File: mm/mm_gran/mm_gran.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+#include "mm_gran/mm_gran.h"
+#include "mm_gran/mm_grantable.h"
+
+#ifdef CONFIG_GRAN
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void gran_free(GRAN_HANDLE handle, FAR void *memory, size_t size)
```
Comment (xiaoxiang781216) [inline mm/mm_gran/mm_gran.c:41]: Can we keep the original file layout and make the change as small as possible? The change is unnecessary big and complex to remove the size limitation.

File: mm/mm_gran/mm_granalloc.c [hunk 3]
```diff
@@ -25,226 +25,46 @@
...
+  int posi;
+  int ret;
+  uintptr_t retp;
+
+  DEBUGASSERT(gran);
+  ngran = NGRANULE(gran, size);
+  if (!ngran || ngran > gran->ngranules) return NULL;
+
+  ret = gran_enter_critical(gran);
+  if (ret < 0)    return NULL;
```
Comment (acassis) [inline mm/mm_gran/mm_granalloc.c:54]: Did you run checkpatch? This kind of coding style is forbidden, it needs use use bracket as in MISRA-C

File: mm/mm_gran/mm_granalloc.c [hunk 4]
```diff
@@ -25,226 +25,46 @@
...
+
+  posi = gran_search(gran, ngran);
+  if (posi >= 0) gran_set(gran, posi, ngran);
```
Comment (acassis) [inline mm/mm_gran/mm_granalloc.c:57]: Ditto

File: mm/mm_gran/mm_granalloc.c [hunk 5]
```diff
@@ -25,226 +25,46 @@
...
+
+  gran_leave_critical(gran);
+  if (posi < 0)  return NULL;
```
Comment (acassis) [inline mm/mm_gran/mm_granalloc.c:60]: Ditto

File: mm/mm_gran/mm_granreserve.c [hunk 6]
```diff
@@ -25,78 +25,54 @@
...
+  ngran = ((end - start) >> gran->log2gran) + 1;
 
-      /* Must lock the granule allocator */
+  /* lock the granule allocator */
 
-      if (gran_enter_critical(priv) < 0)
-        {
-          return NULL;
-        }
+  if (gran_enter_critical(gran) < 0) return NULL;
```
Comment (acassis) [inline mm/mm_gran/mm_granreserve.c:66]: Ditto

File: mm/mm_gran/mm_granreserve.c [hunk 7]
```diff
@@ -25,78 +25,54 @@
...
 
-      /* And reserve the granules */
+  avail = gran_match(gran, posi, ngran, 0, NULL);
+  if (avail) gran_set(gran, posi, ngran);
```
Comment (acassis) [inline mm/mm_gran/mm_granreserve.c:69]: Ditto

File: mm/mm_gran/mm_grantable.c [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/* prepare given GAT range instance for later use. */
+
+int gran_range(const gran_t *gran, size_t posi, size_t size,
+               gatr_t *rang)
+{
+  if (!gran || gran->ngranules < size || !rang) return -EINVAL;
```
Comment (acassis) [inline mm/mm_gran/mm_grantable.c:124]: Ditto

File: mm/mm_gran/mm_grantable.c [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+/* set or clear a range of GAT bits */
+
+static void gran_set_(gran_t *gran, gatr_t *rang, bool val)
+{
+  uint32_t c = rang->sidx + 1;
+
+  cell_set(gran, rang->sidx, rang->smask, val);
+  if (rang->sidx == rang->eidx) return;
```
Comment (acassis) [inline mm/mm_gran/mm_grantable.c:106]: Ditto

**Comment by acassis** [comment]
@xiaoxiang781216 @btashton @pkarashchenko Seems like there is a fail in our nxstyle coding check, it is allowing single line if () without { }, as requested: https://nuttx.apache.org/docs/latest/contributing/coding_style.html

Single line if () without { } is ticking bomb waiting for some victim!

File: mm/mm_gran/mm_granalloc.c (hunk 3 above)
Comment (yf13) [reply mm/mm_gran/mm_granalloc.c:54]: Yes I did. but I will add brackets though I dislike long files.

**Comment by acassis** [comment]
> @xiaoxiang781216 @btashton @pkarashchenko Seems like there is a fail in our nxstyle coding check, it is allowing single line if () without { }, as requested: https://nuttx.apache.org/docs/latest/contributing/coding_style.html
> 
> Single line if () without { } is ticking bomb waiting for some victim!

Strange there is a rule for it in the tools/uncrustify.cfg that Greg and Bob Feretich was trying to do in 2018: 
```
mod_full_brace_if             = add      # "if (a) { a--; }" vs "if (a) a--;"
```

But there isn't a check for it on nxstyle, I noticed we have many files with this issue, I will open an issue and submit a patch later. See:

```
$ git grep "if (" | grep "\;"
```

File: mm/mm_gran/mm_granalloc.c (hunk 3 above)
Comment (acassis) [reply mm/mm_gran/mm_granalloc.c:54]: Far enough! :-D

Single lines with brackets are safe!

**Review by acassis — APPROVED** [review]

**Comment by yf13** [comment]
@acassis thanks for mentioning the uncrustify.cfg thing! The tool can save us a lot code formatting time.
However, it seems that the `tools/uncrustify.cfg` is wrong about block comments format?

**Review by jerpelea — APPROVED** [review]

**Comment by acassis** [comment]
> @acassis thanks for mentioning the uncrustify.cfg thing! The tool can save us a lot time. However, it seems that the latest version of `tools/uncrustify.cfg` is wrong about block comments format?

@yf13 AFAIK Greg and others tried to use uncrustify to check the NuttX coding style, but seems like the NuttX coding style rules were too complex to be able to get catch by this tool, so they gave up. Maybe there are some way to create custom rules checks and get it working. So if it works we could get rid of our nxstyle to use something more standard.

**Comment by yf13** [comment]
@xiaoxiang781216 the `mm_gran.c` has been split into three files as you requested. both patch description and commit message are updated with explanations about the changes. please give a look and let me know how to proceed.

**Pull request merged**